message.go

```diff
@@ -31,16 +36,68 @@ func (msg Message) item() messageSetItem {
 }
 
 func (msg Message) message() message {
+	var attrs int8
+	if msg.CompressionCodec != nil {
+		attrs = int8(msg.CompressionCodec.Code()) & compressionCodecMask
+	}
+
 	m := message{
-		MagicByte: 1,
-		Key:       msg.Key,
-		Value:     msg.Value,
-		Timestamp: timestamp(msg.Time),
+		MagicByte:  1,
+		Key:        msg.Key,
+		Value:      msg.Value,
+		Timestamp:  timestamp(msg.Time),
+		Attributes: attrs,
 	}
 	m.CRC = m.crc32()
 	return m
 }
 
+func (msg Message) encode() (Message, error) {
+	//if msg.CompressionCodec.Code() == CompressionNoneCode {
+	if msg.CompressionCodec == nil {
+		return msg, nil
+	}
+
+	codec, ok := codecs[msg.CompressionCodec.Code()]
+	if !ok {
+		return msg, fmt.Errorf("codec %d not imported.", msg.CompressionCodec.Code())
+	}
+
+	var err error
+	msg.Value, err = transform(msg.Value, codec.Encode)
+	return msg, err
+}
+
+func (msg Message) decode() (Message, error) {
+	c := msg.message().Attributes & compressionCodecMask
+	if c == CompressionNoneCode {
+		return msg, nil
+	}
+
+	codec, ok := codecs[c]
+	if !ok {
+		return msg, fmt.Errorf("codec %d not imported.", msg.CompressionCodec)
```

Comment:
I wonder if an error is the right thing to do here. Maybe we just want to let the message go through in a compressed format?

The challenge is to detect whether the message is compressed or not at the application level then. We could introduce a "UnknownCompressionCodec" implementation of of the `CompressionCodec` interface that programs can test against, for example:
```go
if _, unknown := msg.CompressionCodec.(kafka.UnknownCompressionCodec); unknown {
   // maybe do something?
}
```

Let me know what you think? (I'm also OK if we make this part of a follow up PR)